mimxrt: Provide the machine.RTC module.

robert-hh opened the pull request:
Initial version, using the LP rtc clock.
It provides setting the Date & Time with rtc.init() or rtc.datetime()
and reading date & time with rtc.datetime() or rtc.now().

If no battery is supplied, the default boot time/date is 1970/1/1 0:0:0.
With a battery, the clock continues to run even when the board is not
powered. The clock is quite precise. If not, using rtc.calibration()
may help.

Possible extension: set an alarm with alarm handler. For short periods
< 70 minutes this is also possible with the timer module,
for longer periods this may have to be done.

robert-hh committed:
mimxrt: Provide a RTC module.

Initial version, using the LP rtc clock.
It provides setting the Date & Time with rtc.init() or rtc.datetime()
and reading date & time with rtc.datetime() or rtc.now().

If not battery is supplied, the default boot time/date is 1970/1/1 0:0:0.
With a battery, the clock continues to run even when the board is not
powered. The clock is quite precise. If not, using rtc.calibration()
may help.

Possible extension: set an alarm with alarm handler. For short periods
< 70 minutes this is also possible with the timer module,
for longer periods this may have to be done.

dpgeorge commented:
Thanks for this.  But we really need to first standardise on a format for the RTC tuple.  It should really match CPython's `datetime.datetime`.  What do you implement here?

robert-hh commented:
I used the format documented under machine.RTC(). So it's (year, month, day, weekday, hour, minute, second, fraction). I can change that anytime, if there is a standard.

robert-hh commented:
I just looked at the Python datetime.datetime() format. That 's obviously easy. The built-in RTC does not support weekday, and I only implemented that for consistency with the MP docs. tzinfo could be None in the first step. But RTC could still be useful in combination with the file system, especially since MIMXRT MCU's support battery backup for the RTC (and it works).

robert-hh committed:
mimxrt: Change in datetime tuple order.

The tuple order for datetime was changed to the CPython sequence:
Year, Month, Day, Hour, Minute, Second, µsecond, TZ
TZ is ignored and reported as None.
µsecond is used at best effort.

The method weekday() was added to report the weekday of the current
date. It starts with 0 for Monday.

robert-hh committed:
mimxrt: Remove _WFE() from MICROPY_EVENT_POLL_HOOK.

dpgeorge commented:
Merged in 2f365d234e9e985710a174c05091054c011e37d7 with a fix for setting the datetime, to use the new tuple ordering/format.
